### ISSUE TITLE
Suggestion: Replace username/password fields with “Isolation Parameter” for TOR type proxies

### ISSUE THREAD

**Issue by ntninja** [foxyproxy/browser-extension#77]
In the context of TOR, the sent username and password fields are not used for authentication (as their name would imply) but instead cause the TOR daemon to use independently established network paths for each username/password combination it receives. When combined with other client isolation features (such as Firefox Containers), this allows visiting the same website in a way that makes it very hard for the remote site to know that the requests originated from the same client.

Since FoxyProxy now already distinguishes between SOCKS5 and TOR in its UI, I’d think its not unreasonable also change the presentation of the username and password fields to something that aligns with their actual meaning in this context: a single field allowing to set an “context identifier”.

Implementation-wise this would probably just be the username-field, but renamed, with the password-field hidden and being set to the same value as the username-field on save.

**Comment by erosman** [comment]
TOR is regarded as SOCKS5. 
The user/pass are intended for SOKS5.
The interface fills some entries to help the user, they are editable.

**Comment by ntninja** [comment]
Yes, the suggestion was just to make the user interface match the meaning of these fields in the context of the user-selected TOR option rather than what they would mean for a generic SOCKS5 proxy. I mean, the way TOR reuses the authentication fields for its stream isolation features has absolutely nothing to do with authentication, yet users end up seeing a username and password field anyways but it doesn’t do what it says on the tin can.

To be clear: What FoxyProxy does here isn’t wrong, the suggestion was just that *since FoxyProxy* already distinguishes between TOR and other SOCKS5 in the UI, I’d be nice if all the fields present in the UI also reflected their meaning in the context of TOR (rather than generic SOCKS5) if that was selected.

**Comment by erosman** [comment]
TBH, changing the UI requires more work than filling or hiding text boxes.
The user/pass boxes represent the data that is sent by the [proxy.ProxyInfo](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/proxy/ProxyInfo) API.
What TOR does with the data that is sent as username & password is beyond the scope of the API.

**Comment by erosman** [comment]
It seems TOR & TOR Browser support username/password

- [My internet connection requires an HTTP or SOCKS Proxy](https://support.torproject.org/tbb/tbb-47)

```
  Socks5Proxy 10.0.0.1:1080
  Socks5ProxyUsername myuser
  Socks5ProxyPassword mypass
```

- [SOCKS5 Proxy Setup Guide for Tor Browser](https://intercom.help/privatevpn/en/articles/6366326-socks5-proxy-setup-guide-for-tor-browser)

![image](https://github.com/foxyproxy/browser-extension/assets/6458876/9d69fb11-f3ff-4145-9160-3d515512fcb7)

**Comment by ntninja** [comment]
@erosman: That documentation is for how the TOR node (or TOR browser with its internal TOR node) connects to the Internet, not for how your browser connects to the TOR node. Apparently TOR browser reuses the Firefox proxy settings for this, but that doesn’t mean anything since all connections from TOR browser go through its TOR node first before they go anywhere else (anything else is considered a serious bug by the devs there).

That VPN provider you linked uses this to easily allow setting up a TOR over VPN setup that some people like to use to hide the fact that they use TOR from their ISP. (Probably stupid in most cases, but VPN providers are very creative at convincing their users that they’re useful…) It has nothing to do how Firefox/Chromium with FoxyProxy or the TOR browser will connect to the locally running TOR node.

**Comment by erosman** [comment]
Is that to say TOR never uses username/password?

I am not sure how Firefox sends these data via socks. 
The API calls them username/password.
https://searchfox.org/mozilla-central/source/toolkit/components/extensions/ProxyChannelFilter.sys.mjs#52-71

**Comment by ntninja** [comment]
The TOR NODE only *accepts* a username/password for stream isolation, it may however be configured to *send* a username/password to an upstream Socks/HTTP proxy. The upstream Socks/HTTP proxy used by the TOR NODE is not visible or configured by FoxyProxy. The username/password entered into FoxyProxy is the one used by the TOR NODE for stream isolation.

The TOR BROWSER on the other hand only offers the option to configure the TOR NODE to *send* a username/password configuration to an upstream Socks/HTTP proxy, not the username/password combination sent to its internal TOR NODE, from what I can tell. At least that this definitely what your screenshot/guide is talking about.

I am not sure what setting the username/password would do when FoxyProxy is installed into the TOR BROWSER or whether FoxyProxy will work in the TOR BROWSER at all.

Note also that TOR BROWSER uses patched FIREFOX code. What you find in SearchFox may be different from the TOR BROWSER is using.

**Comment by erosman** [comment]
I have been looking for details on this issue. Where exactly is it documented?

**Comment by ntninja** [comment]
Documentation for this is available here:
https://spec.torproject.org/socks-extensions.html#extended-auth

Apparently what I had described in 2023 is called *legacy isolation* – not sure if that was already legacy back then, but probably…

The updated protocol is to set username to `<torS0X>0` (stream isolation without RPC Object ID[1]) and the password to the stream isolation field.

[1]: RPC Object IDs can only be allocated by [using TORs IPC protocol](https://people.torproject.org/~nickm/volatile/rpc-reference.html#method:arti:new_isolated_client) (which Firefox doesn’t support), so that feature isn’t really useful, I think.

**Comment by erosman** [comment]
Is that to say that using [usernames and passwords](https://spec.torproject.org/socks-extensions.html#extended-auth) would be fine in the user-interface?

**Comment by ntninja** [comment]
@erosman: No, it should only be the stream context name (displayed as visible text field, but sent as SOCKS5 password) with the SOCKS5 username being hard-coded to `<torS0X>0`.

**Comment by erosman** [comment]
> I have been looking for details on this issue. Where exactly is it documented?

Where does it say that TOR uses “Context Identifier” (or any other term) and why is "username" inappropriate?

**Comment by ntninja** [comment]
@erosman: Well, the document calls it the “stream isolation parameter” (or perhaps just “isolation parameter” – a STREAM in TOR lingue is a single connection through the TOR network by a TOR client). The term “Context Identifier” doesn’t actually seem to be used by the TOR project, you’re right about that! (I cannot recall where I got it from, but probably best to drop that term then!)

Calling it “username” is inappropriate however, since the isolation parameter has nothing to do with authentication in any sense, it merely reuses the SOCKS authenticate mechanism for broader client compatibility than any custom protocol would allow.

**Comment by erosman** [comment]
> Calling it “username” is inappropriate however, since the isolation parameter has nothing to do with authentication in any sense, it merely reuses the SOCKS authenticate mechanism for broader client compatibility than any custom protocol would allow.

Firefox sends it as username and FoxyProxy sets the value according to the Firefox API.
How TOR interprets the data sent as username, is a up to them.

https://searchfox.org/mozilla-central/source/toolkit/components/extensions/ProxyChannelFilter.sys.mjs#231-243

https://github.com/foxyproxy/browser-extension/blob/90ea71e185589f37f91e2333bfa0175a79f619b3/src/content/on-request.js#L62-L70

**Comment by ntninja** [comment]
@erosman: I think you are confused be the proposal here! (Not that the back and forth and updates didn’t make it confusing of course! 😉)

As things are, I can already make use of TOR’s stream isolation feature by manually configuring it like this:

![TOR proxy settings section with username set to “<torS0X>0” and password set to “Firefox/Incognito”](https://github.com/user-attachments/assets/58aad06a-c254-4db5-9fa3-30aa4b37dcf5)

This is not at all intuitive however, since the “username” is not actually a username but a request format type and the “password” is not actually a password but the isolation parameter. The proposal, hence, is to replace these two fields in the UI with a properly labelled one that reads “Stream Isolation” (or similar) – ideally with a link or short summary of what it does. That fact that this stream isolation parameter is then **encoded** as SOCKS authentication data is really an implementation detail for the layers below. (Also country/city should be grayed out since TOR doesn’t actually allow selecting that per-request. – It’s always just “somewhere”.)

**Comment by erosman** [comment]
Firefox _(and thus FoxyProxy)_ sends the data as `username`.

What the 3rd party server at the receiving end calls the data sent as `username`, is their choice _(e.g. 'registration id', 'customer number', etc)_ and beyond the scope of the extension. Altering the user-interface _(processing, translations, etc)_ for what each 3rd party server likes to call it, is not feasible.

The Server list in the UI is simply an auto-fill to help the users in filling a few input entries.

**Comment by ericjung** [comment]
Hello @ntninja. @erosman and I agreed not to change the name of the field (username ==> isolation parameter). However, we are willing to set the VALUE of the username to `<torS0X>0` when TOR is selected from the 'Type' list. Can you confirm this should always be `<torS0X>0`? Last time I used the SOCKS5 proxy server on 127.0.0.1:9050 that comes with TOR this value was supposed to be empty/blank.

**Comment by ntninja** [comment]
> Hello [@ntninja](https://github.com/ntninja). [@erosman](https://github.com/erosman) and I agreed not to change the name of the field (username ==> isolation parameter). However, we are willing to set the VALUE of the username to `<torS0X>0` when TOR is selected from the 'Type' list. Can you confirm this should always be `<torS0X>0`? Last time I used the SOCKS5 proxy server on 127.0.0.1:9050 that comes with TOR this value was supposed to be empty/blank.

As far as I know there is no difference whether the username is `<torS0X>0` or empty **iff** the password field is also empty. So setting it to `<torS0X>0` wouldn’t break anything and at least hint at the fact that TOR treats the SOCKS5 username field as a request type rather than any form of user identifier (and notably does not, as @erosman implied, treat it as *'registration id', 'customer number', etc*!).

(Also to highlight again, since it didn’t seem clear: It’s the SOCKS5 password field that is treated as the “isolation parameter” – and is ideally displayed without password field obfuscation/stars in the UI. The SOCKS5 username should always either be the mentioned `<torS0X>0` or empty unless one is using special features of the TOR API.)

**Comment by erosman** [comment]
I would imagine, people who are using TOR would understand how to use the input fields.

Nonetheless, if there is a popular demand for this feature, it will be considered.
